tests/ui/associated-consts/defaults-not-assumed-fail.stderr

```diff
@@ -10,6 +10,14 @@ note: erroneous constant encountered
 LL |     assert_eq!(<() as Tr>::B, 0);    // causes the error above
    |                ^^^^^^^^^^^^^
 
+note: erroneous constant encountered
+  --> $DIR/defaults-not-assumed-fail.rs:33:16
+   |
+LL |     assert_eq!(<() as Tr>::B, 0);    // causes the error above
+   |                ^^^^^^^^^^^^^
```

Comment:
I think we should figure out how to avoid this duplication. We spent a bunch of work a few years back to cut down on the duplicates, it'd be a shame to lose all that again.

The first step would be to figure out why we even get duplicates. Does promotion leave copies of the MIR in the old body? If yes we probably want to either run this pass before promotion (IIRC @oli-obk tried that and it had major perf issues, but maybe those can be figured out), or we'll want to do some basic dead code analysis before const-prop-lint.

Reply:
The fact that https://github.com/rust-lang/rust/issues/117949 is an issue indicates that we *do* remove the original operation from the MIR. So then the question is, why do we still get duplicates?

Reply:
Ah... I think for shifts we remove the original operation, since all it does is compute the shift. But for addtion, we actually end up duplicating the `CheckedAdd`, it is once in the promoted (to compute the result) and once in the original MIR (to compute whether the assertion failed).

See [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=559267a3939e9717d7f644091e28b53f):
```
fn foo() -> &i32 {
    let mut _0: &i32;
    let mut _1: (i32, bool);

    bb0: {
        _1 = CheckedAdd(const 1_i32, const 2_i32);
        assert(!move (_1.1: bool), "attempt to compute `{} + {}`, which would overflow", const 1_i32, const 2_i32) -> [success: bb1, unwind continue];
    }

    bb1: {
        _0 = const _;
        return;
    }
}

promoted[0] in foo: &i32 = {
    let mut _0: &i32;
    let mut _1: i32;
    let mut _2: (i32, bool);

    bb0: {
        _2 = CheckedAdd(const 1_i32, const 2_i32);
        _1 = move (_2.0: i32);
        _0 = &_1;
        return;
    }
}
```

That seems rather silly. I wonder if we can somehow avoid that duplication. Or at least mark one of these `CheckedAdd` in a way that const-prop-lint can ignore them.

Reply:
considering it's not shown to users (it's an exact duplicate after all), it's not too important, right?

Reply:
>  we probably want to [...] run this pass before promotion 

That is the most logical solution - run the lint when the MIR most closely matches the user's code. It will catch all the issues and produce zero duplicates. However, I [tried that and it lead to a cycle](https://github.com/rust-lang/rust/issues/117949#issuecomment-1871194279) so I abandoned the idea.

> Ah... I think for shifts we remove the original operation, since all it does is compute the shift. But for addtion, we actually end up duplicating the CheckedAdd, it is once in the promoted (to compute the result) and once in the original MIR (to compute whether the assertion failed).

Yes, only the shifts are not duplicated. All other ops are.

> considering it's not shown to users (it's an exact duplicate after all), it's not too important, right?

Yeah, I feel so. We can come back and fix the duplicates in a later PR.

Reply:
> we probably want to [...] run this pass before promotion

That is the most logical solution - run the lint when the MIR most closely matches the user's code. It will catch all the issues and produce zero duplicates. However, I https://github.com/rust-lang/rust/issues/117949#issuecomment-1871194279 so I abandoned the idea.

> Ah... I think for shifts we remove the original operation, since all it does is compute the shift. But for addtion, we actually end up duplicating the CheckedAdd, it is once in the promoted (to compute the result) and once in the original MIR (to compute whether the assertion failed).

Yes, only the shifts are not duplicated. All other ops are.

> considering it's not shown to users (it's an exact duplicate after all), it's not too important, right?

Yeah, I feel so. We can come back and fix the duplicates in a later PR.

Reply:
Fair, but let's open an issue to track cleaning this up.

Reply:
Created issue #119647